File: app/src/main/java/com/pkfcooperparry/template/data/local/db/AppDbHelper.java
package com.pkfcooperparry.template.data.local.db;

import com.pkfcooperparry.template.data.model.db.User;
import com.pkfcooperparry.template.data.model.db.visit.Form;
import com.pkfcooperparry.template.data.model.db.visit.Page;
import com.pkfcooperparry.template.data.model.db.visit.Tab;

import java.util.List;

import javax.inject.Inject;
import javax.inject.Singleton;

import io.reactivex.Observable;
import io.realm.Realm;

/**
 * The class will contain methods to access REALM and return or create data
 */

@Singleton
public class AppDbHelper implements DbHelper {

    private final Realm realm;

    @Inject
    public AppDbHelper(Realm appDatabase) {
        this.realm = appDatabase;
    }

    @Override
    public Observable<List<User>> getAllUsers() {
        return Observable.fromCallable(() -> realm.where(User.class).findAll());
    }

    @Override
    public List<Form> getForms() {
        return realm.where(Form.class).findAll();
    }


    @Override
    public User getUser() {
        return realm.where(User.class).findFirst();
    }

    @Override
    public void addUser(final User user) {
        realm.executeTransaction(realm -> realm.copyToRealmOrUpdate(user));
    }

    @Override
    public void updateUser(final User user, final String firstName, final String surname, final int sex, final int age, final String hometown, final int positionId, final String phone) {
        realm.executeTransaction(realm -> {
            user.setName(firstName);
            user.setSurname(surname);
            user.setAge(sex);
            user.setGender(sex);
            user.setAge(age);
            user.setHomeTown(hometown);
            user.setUsualPosition(positionId);
            user.setPhone(phone);
            realm.copyToRealmOrUpdate(user);
        });
    }

    @Override
    public void deleteAll() {
        realm.executeTransaction(realm -> realm.deleteAll());
    }

    @Override
    public Tab getTab(long tabId) {
        return realm.where(Tab.class).equalTo("tabId", tabId).findFirst();
    }

    @Override
    public Form getFormWithId(int formId) {
        return realm.where(Form.class).equalTo("id", formId).findFirst();

    }

    @Override
    public Page getPageWithId(int pageId) {
        return realm.where(Page.class).equalTo("pageId", pageId).findFirst();

    }


    @Override
    public void addForms(List<Form> forms) {
        realm.executeTransaction(realm -> {
            realm.copyToRealmOrUpdate(forms);
            for (Form form : forms) {
                realm.copyToRealmOrUpdate(form.getPages());

                for (Page page : form.getPages()) {
                    realm.copyToRealmOrUpdate(page.getTabs());

                }
            }
        });
    }


}

File: app/src/main/java/com/televantou/template/data/local/db/AppDbHelper.kt
package com.televantou.template.data.local.db

import com.televantou.template.data.model.db.User
import com.televantou.template.data.model.db.visit.Form
import com.televantou.template.data.model.db.visit.Page
import com.televantou.template.data.model.db.visit.Tab
import io.reactivex.Observable
import io.realm.Realm
import javax.inject.Inject
import javax.inject.Singleton

/**
 * The class will contain methods to access REALM and return or create data
 */

@Singleton
class AppDbHelper @Inject
constructor(private val realm: Realm) : DbHelper {

    override val allUsers: Observable<List<User>>
        get() = Observable.fromCallable { realm.where(User::class.java).findAll() }

    override val forms: List<Form>
        get() = realm.where(Form::class.java).findAll()


    override val user: User?
        get() = realm.where(User::class.java).findFirst()

    override fun addUser(user: User) {
        realm.executeTransaction { realm -> realm.copyToRealmOrUpdate(user) }
    }

    override fun updateUser(user: User, firstName: String, surname: String, sex: Int, age: Int, hometown: String, positionId: Int, phone: String) {
        realm.executeTransaction { realm ->
            user.name = firstName
            user.surname = surname
            user.age = sex
            user.gender = sex
            user.age = age
            user.homeTown = hometown
            user.usualPosition = positionId
            user.phone = phone
            realm.copyToRealmOrUpdate(user)
        }
    }

    override fun deleteAll() {
        realm.executeTransaction { realm -> realm.deleteAll() }
    }

    override fun getTab(tabId: Long): Tab? {
        return realm.where(Tab::class.java).equalTo("tabId", tabId).findFirst()
    }

    override fun getFormWithId(formId: Int): Form? {
        return realm.where(Form::class.java).equalTo("id", formId).findFirst()

    }

    override fun getPageWithId(pageId: Int): Page? {
        return realm.where(Page::class.java).equalTo("pageId", pageId).findFirst()

    }


    override fun addForms(forms: List<Form>) {
        realm.executeTransaction { realm ->
            realm.copyToRealmOrUpdate(forms)
            for (form in forms) {
                realm.copyToRealmOrUpdate(form.pages)

                for (page in form.pages!!) {
                    realm.copyToRealmOrUpdate(page.tabs)

                }
            }
        }
    }


}
